src/H5PB.c

```diff
@@ -208,1370 +237,4629 @@ H5PB_get_stats(const H5PB_t *page_buf, unsigned accesses[2], unsigned hits[2], u
+ *
+ *                    and proceed to the next index entry that doesn't
+ *                    appear in the tick list.
+ *
+ *                 3) If the cognate entry does appear in the page buffer
+ *                    and is clean, proceed as per 2) above.
+ *
+ *                 4) In all other cases, do nothing, and proceed to the
+ *                    next index entry that does not appear in the tick list.
+ *
+ *
+ * Return:	Non-negative on success/Negative on failure
+ *
+ * Programmer:	John Mainzer -- 11/9/18
+ *
+ * Changes:     None.
+ *
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5PB_vfd_swmr__update_index(H5F_t *f, uint32_t *idx_ent_added_ptr, uint32_t *idx_ent_modified_ptr,
+                            uint32_t *idx_ent_not_in_tl_ptr, uint32_t *idx_ent_not_in_tl_flushed_ptr)
+{
+    H5F_shared_t *const        shared   = f->shared;
+    const uint64_t             tick_num = shared->tick_num;
+    uint32_t                   i;
+    uint32_t                   idx_ent_added             = 0;
+    uint32_t                   idx_ent_modified          = 0;
+    uint32_t                   idx_ent_not_in_tl         = 0;
+    uint32_t                   idx_ent_not_in_tl_flushed = 0;
+    H5PB_t *                   page_buf                  = NULL;
+    H5PB_entry_t *             entry;
+    H5FD_vfd_swmr_idx_entry_t *ie_ptr    = NULL;
+    H5FD_vfd_swmr_idx_entry_t *idx       = NULL;
+    herr_t                     ret_value = SUCCEED; /* Return value */
+
+    FUNC_ENTER_NOAPI(FAIL)
+
+    HDassert(shared->vfd_swmr);
+    HDassert(shared->vfd_swmr_writer);
+
+    idx = shared->mdf_idx;
+
+    HDassert(idx);
+
+    page_buf = shared->page_buf;
+
+    HDassert(page_buf);
+    HDassert(page_buf->magic == H5PB__H5PB_T_MAGIC);
+    HDassert(page_buf->vfd_swmr_writer);
+
+    HDassert(idx_ent_added_ptr);
+    HDassert(idx_ent_modified_ptr);
+    HDassert(idx_ent_not_in_tl_ptr);
+    HDassert(idx_ent_not_in_tl_flushed_ptr);
+
+    /* scan the tick list and insert or update metadata file index entries
+     * as appropriate.
+     */
+
+    for (entry = page_buf->tl_head_ptr; entry != NULL; entry = entry->tl_next) {
+        uint64_t target_page = entry->page;
+
+        HDassert(entry->magic == H5PB__H5PB_ENTRY_T_MAGIC);
+
+        /* see if the shadow index already contains an entry for *entry. */
+
+        ie_ptr = H5FD_vfd_swmr_pageno_to_mdf_idx_entry(idx, shared->mdf_idx_entries_used, target_page, FALSE);
+
+        if (ie_ptr == NULL) { /* alloc new entry in the metadata file index*/
+            uint32_t new_index_entry_index;
+
+            new_index_entry_index = shared->mdf_idx_entries_used + idx_ent_added++;
+
+            if (new_index_entry_index >= shared->mdf_idx_len &&
+                (idx = H5F_vfd_swmr_enlarge_shadow_index(f)) == NULL) {
+                HDfprintf(stderr, "\n\nmax mdf index len (%" PRIu32 ") exceeded.\n\n", shared->mdf_idx_len);
```

Comment:
Exiting out of the middle of the library is not allowed.   This must be corrected.